src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceActivationMaintainer.cs

```diff
@@ -0,0 +1,387 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.GrainDirectory;
+using Orleans.SystemTargetInterfaces;
+using Orleans.Runtime.Configuration;
+using Orleans.Runtime.Scheduler;
+
+namespace Orleans.Runtime.GrainDirectory
+{
+    internal class GlobalSingleInstanceActivationMaintainer : AsynchAgent
+    {
+        private LocalGrainDirectory router;
+        private Logger logger;
+        private TimeSpan period;
+
+        internal GlobalSingleInstanceActivationMaintainer(LocalGrainDirectory router, Logger logger, GlobalConfiguration config)
+        {
+            this.router = router;
+            this.logger = logger;
+            this.period = config.GlobalSingleInstanceRetryInterval;
+            logger.Verbose("GSIP:M GlobalSingleInstanceActivationMaintainer Started, Period = {0}", period);
+
+        }
+
+        // scanning the entire directory for doubtful activations is too slow.
+        // therefore, we maintain a list of potentially doubtful activations on the side.
+        // maintainer periodically takes and processes this list.
+        private List<GrainId> doubtfulGrains = new List<GrainId>();
+        private object lockable = new object();
+
+        public void TrackDoubtfulGrain(GrainId grain)
+        {
+            lock (lockable)
+                doubtfulGrains.Add(grain);
+        }
+        public void TrackDoubtfulGrains(Dictionary<GrainId, IGrainInfo> newstuff)
+        {
+            var newdoubtful = newstuff.Where(kp =>
+             {
+                 if (!kp.Value.SingleInstance) return false;
+                 var act = kp.Value.Instances.FirstOrDefault();
+                 if (act.Key == null) return false;
+                 if (act.Value.RegistrationStatus == MultiClusterStatus.Doubtful) return true;
+                 return false;
+             }).Select(kp => kp.Key);
```

Comment:
I think you should add a `ToList()` here so the Linq query will be executed now, and we will spend less time in the lock statement below. 

And very minor but you could write this in pure Linq (no if/return in the where method), I find it more readable, but maybe it is just me.